src/controller/event/handler/webhook/artifact/retention.go

```diff
@@ -0,0 +1,144 @@
+package artifact
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/goharbor/harbor/src/core/api"
+
+	"github.com/goharbor/harbor/src/controller/event"
+	"github.com/goharbor/harbor/src/controller/event/handler/util"
+	"github.com/goharbor/harbor/src/core/config"
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/goharbor/harbor/src/pkg/notification"
+	"github.com/goharbor/harbor/src/pkg/notifier/model"
+)
+
+// RetentionHandler preprocess tag retention event data
+type RetentionHandler struct {
+}
+
+// Handle ...
+func (r *RetentionHandler) Handle(value interface{}) error {
+	if !config.NotificationEnable() {
+		log.Debug("notification feature is not enabled")
+		return nil
+	}
+
+	trEvent, ok := value.(*event.RetentionEvent)
+	if !ok {
+		return errors.New("invalid tag retention event type")
+	}
+	if trEvent == nil {
+		return fmt.Errorf("nil tag retention event")
+	}
+
+	payload, dryRun, project, err := constructRetentionPayload(trEvent)
+	// if dry run, do not trigger webhook
+	if dryRun {
+		log.Debugf("retention task %v is dry run", trEvent.TaskID)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	policies, err := notification.PolicyMgr.GetRelatedPolices(project, trEvent.EventType)
+	if err != nil {
+		log.Errorf("failed to find policy for %s event: %v", trEvent.EventType, err)
+		return err
+	}
+	if len(policies) == 0 {
+		log.Debugf("cannot find policy for %s event: %v", trEvent.EventType, trEvent)
+		return nil
+	}
+	err = util.SendHookWithPolicies(policies, payload, trEvent.EventType)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// IsStateful ...
+func (r *RetentionHandler) IsStateful() bool {
+	return false
+}
+
+func constructRetentionPayload(event *event.RetentionEvent) (*model.Payload, bool, int64, error) {
+	task, err := api.RetentionController.GetRetentionExecTask(event.TaskID)
+	if err != nil {
+		log.Errorf("failed to get retention task %d: error: %v", event.TaskID, err)
+		return nil, false, 0, err
+	}
+	if task == nil {
+		return nil, false, 0, fmt.Errorf("task %d not found with retention event", event.TaskID)
+	}
+
+	execution, err := api.RetentionController.GetRetentionExec(task.ExecutionID)
+	if err != nil {
+		log.Errorf("failed to get retention execution %d: error: %v", task.ExecutionID, err)
+		return nil, false, 0, err
+	}
+	if execution == nil {
+		return nil, false, 0, fmt.Errorf("execution %d not found with retention event", task.ExecutionID)
+	}
+
+	if execution.DryRun {
+		return nil, true, 0, nil
+	}
+
+	md, err := api.RetentionController.GetRetention(execution.PolicyID)
+	if err != nil {
+		log.Errorf("failed to get tag retention policy %d: error: %v", execution.PolicyID, err)
+		return nil, false, 0, err
+	}
+	if md == nil {
+		return nil, false, 0, fmt.Errorf("policy %d not found with tag retention event", execution.PolicyID)
+	}
+
+	extURL, err := config.ExtURL()
+	if err != nil {
+		log.Errorf("Error while reading external endpoint URL: %v", err)
```

Comment:
no return?

Reply:
> no return?

I'm considering that the hostname get from the extURL is not essential info of webhook, so I will keep sending webhook when getting extURL failed.

Reply:
But, can you make sure line 104 can be worked fine as there is an error raised before? It seems that the extURL is still be parsed, isn't it should be parsed only on success to get the extURL?

can we set a default value, like "" or whatever meaningful string, for hostname if an error happens?

Reply:
1. ExtURL() will always return nil error 
2. if extURL is empty, hostname will be empty("")